plugins/inputs/dpdk/dpdk.go

```diff
@@ -0,0 +1,240 @@
+	Log                telegraf.Logger `toml:"-"`
+	connector          *dpdkConnector  `toml:"-"`
+	rawdevCommands     []string        `toml:"-"`
+	ethdevCommands     []string        `toml:"-"`
+}
+
+type ethdevConfig struct {
+	EthdevExcludeCommands []string `toml:"exclude_commands"`
+}
+
+func init() {
+	inputs.Add(dpdkPluginName, func() telegraf.Input {
+		dpdk := &dpdk{
+			SocketPath:     defaultPathToSocket,
+			AccessTimeout:  defaultAccessTimeout,
+			rawdevCommands: []string{"/rawdev/xstats"},
+			ethdevCommands: []string{"/ethdev/stats", "/ethdev/xstats", "/ethdev/link_status"},
+		}
+		return dpdk
+	})
+}
+
+func (dpdk *dpdk) SampleConfig() string {
+	return dpdkSampleConfig
+}
+
+func (dpdk *dpdk) Description() string {
+	return dpdkDescription
+}
+
+// Performs validation of all parameters from configuration
+func (dpdk *dpdk) Init() error {
+	if dpdk.SocketPath == "" {
+		dpdk.SocketPath = defaultPathToSocket
+		dpdk.Log.Infof("using default '%v' path for socket_path", defaultPathToSocket)
+	}
+	if err := isSocket(dpdk.SocketPath); err != nil {
+		return err
+	}
+
+	if err := dpdk.validateCommands(); err != nil {
+		return err
+	}
+
+	if dpdk.AccessTimeout < 0 {
+		return fmt.Errorf("socket_access_timeout should be positive number or equal to 0 (to disable timeouts)")
+	}
+	if len(dpdk.AdditionalCommands) == 0 && len(dpdk.DeviceTypes) == 0 {
+		dpdk.Log.Warn("DPDK plugin is enabled and configured with nothing to read")
+	}
+
+	dpdk.connector = newDpdkConnector(dpdk.SocketPath, dpdk.AccessTimeout)
+	result, err := dpdk.connector.connect()
+	if result != "" {
+		dpdk.Log.Info(result)
+	}
+	return err
+}
+
+// Checks that user-supplied commands are unique and match DPDK commands format
+func (dpdk *dpdk) validateCommands() error {
+	dpdk.AdditionalCommands = uniqueValues(dpdk.AdditionalCommands)
+
+	for _, commandWithParams := range dpdk.AdditionalCommands {
+		if len(commandWithParams) == 0 {
+			return fmt.Errorf("got empty command")
+		}
+
+		if commandWithParams[0] != '/' {
+			return fmt.Errorf("'%v' command should start with '/'", commandWithParams)
+		}
+
+		if commandWithoutParams := stripParams(commandWithParams); len(commandWithoutParams) >= maxCommandLength {
+			return fmt.Errorf("'%v' command is too long. It shall be less than %v characters", commandWithoutParams, maxCommandLength)
+		}
+
+		if len(commandWithParams) >= maxCommandLengthWithParams {
+			return fmt.Errorf("command with parameters '%v' shall be less than %v characters", commandWithParams, maxCommandLengthWithParams)
+		}
+	}
+
+	return nil
+}
+
+// Gathers all unique commands and processes each command sequentially
+// Parallel processing could be achieved by running several instances of this plugin with different settings
+func (dpdk *dpdk) Gather(accumulator telegraf.Accumulator) error {
+	commands := dpdk.gatherCommands(accumulator)
+
+	for _, command := range commands {
+		dpdk.processCommand(command, accumulator)
+	}
```

Comment:
Is this required to be executed sequentially?

Reply:
Connections and timeouts management for parallel approach (with goroutines) would introduce additional complexity we would like to avoid. We tested this approach initially, but this is much easier to understand and work with.

Reply:
Perfectly fine with me. Keep as is and mark as resolved.